core/src/main/scala/org/apache/spark/rdd/RDDCheckpointData.scala

```diff
@@ -44,6 +46,26 @@ private[spark] class RDDCheckpointData[T: ClassTag](@transient rdd: RDD[T])
 
   import CheckpointState._
 
+  // Because SparkContext is transient in RDD, so we can't get the id and checkpointDir later.
+  // So keep a copy of the id and checkpointDir.
```

Comment:
Not sure if I understand this comment. How did it work before then? Even before this patch `doCheckpoint` directly calls `rdd.id`